### PR TITLE
Add alternative tool for incremental builds when tuning

### PR DESCRIPTION
**Pull request by davidd-amd**
During the tuning process, it is of interest to modify an assembly file/s and rebuild the corresponding object file/s and then relink the corresponding co file. Currently, we generate additional source files and a script to provide this workflow (which requires using `MergeFiles: False`). 

This PR adds a new `Makefile` that manages rebuilding a co file during iterative development when tuning. One modifies an assembly file of interest, then runs `make` and make will detect what file/s changed and rebuild accordingly.

Assumptions:

- Each problem directory contains a library directory with one co file corresponding to one architecture

Example:

```
cd hipBLASLt/tensilelite
Tensile/bin/Tensile Tensile/Tests/gemm/fp16_use_e.yaml tensile-out
# modify an assembly file in tensile-out/1_BenchmarkProblems/Cijk_Ailk_Bjlk_DB_UserArgs_00/00_Final/source/build_tmp/SOURCE/assembly
make co TENSILE_OUT=tensile-out
# re-run the client
```

The Makefile will set the target based on the name of the co file and sets a default wavefront flag but each of these can be customized as follows:

```
make co TENSILE_OUT=tensile-out ARCH="gfx942:xnack-" WAVEFRONTSIZE="-mwavefrontsize64"
```

In addition, we provide `ASM_ARGS` and `LINK_ARGS` as additional customization points for the assemble and link step respectively.

### CODE REVIEW

**Comment by davidd-amd** [comment]
@KKyang @jichangjichang Let me know if this Makefile is suitable for your development process.

File: tensilelite/Makefile [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+# ├── 0_Build
+# ├── 1_BenchmarkProblems
+# │   ├── Cijk_Ailk_Bjlk_1
+# |   ├── Cijk_Ailk_Bjlk_2
+# |   ...
+# |   ├── Cijk_Ailk_Bjlk_N
+#
+# One can invoke make with this Makefile as follows:
+#
+# make co PROBLEM_DIR=bar-out/1_BenchmarkProblems/Cijk_Ailk_Bjlk_1
```
Comment (bstefanuk) [inline tensilelite/Makefile]:
```suggestion
# make co PROBLEM_DIR=tensile-out/1_BenchmarkProblems/Cijk_Ailk_Bjlk_1
```
I think this should be "tensile-out" to match the example

Comment (davidd-amd) [reply tensilelite/Makefile]: cp paste...

File: tensilelite/Makefile [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+#
+# where PROBLEM_DIR is the relative path to one problem directory.
+# ------------------------------------------------------------------------
+
+AS := /opt/rocm/bin/amdclang++
+LDD := /opt/rocm/bin/amdclang++
+SRCFILES := $(shell find ${PROBLEM_DIR}/00_Final/source/build_tmp/SOURCE/assembly -name '*.s')
```
Comment (bstefanuk) [inline tensilelite/Makefile]: The directory "SOURCE" in this path can change depending on the build directory specified by TensileCreateLibrary. Is this Makefile only intended for use with TensileLt builds issued via ./install.sh?

**Comment by davidd-amd** [comment]
Adding noci label because this PR adds code that isn't used in testing.

File: tensilelite/Makefile (hunk 2 above)
Comment (davidd-amd) [reply tensilelite/Makefile]: This is intended to be used with `Tensile` not `TensileCreateLibrary`. It also assumes the default directory structure that one gets when running `Tensile`. If there is a way that the user can configure the directory structure such that the Makefile breaks please describe it here.

File: tensilelite/Tensile/TensileCreateLibrary.py [hunk 3]
```diff
@@ -361,65 +361,6 @@ def buildSourceCodeObjectFiles(CxxCompiler, kernelFiles, outputPath):
 
     return itertools.chain.from_iterable(coFiles)
 
-################################################################################
-def prepAsm(kernelWriterAssembly):
```
Comment (KKyang) [inline tensilelite/Tensile/TensileCreateLibrary.py:365]: Is this an unused function?

Comment (davidd-amd) [reply tensilelite/Tensile/TensileCreateLibrary.py:365]: That function generates the asm-new.sh script that is intended to be replaced by the Makefile added in this PR.

**Comment by davidd-amd** [comment]
Removing noCI label as we removed existing capability.

**Review by KKyang — DISMISSED** [review]

File: tensilelite/Makefile [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+# the library directory and sets a default wavefront flag but each of 
+# these can be customized as follows:
+#
+# make co TENSILE_OUT=tensile-out ARCH="gfx942:xnack-" WAVEFRONTSIZE="-mno-wavefrontsize64"
+#
+# In addition, we provide AS and ASM_ARGS, and LD LINK_ARGS as customization
+# points for the assemble and link step respectively.
+# ------------------------------------------------------------------------
+
+AS := /opt/rocm/bin/amdclang++
```
Comment (jichangjichang) [inline tensilelite/Makefile]: Should we check ROCM_PATH?

File: tensilelite/Makefile [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+.SECONDEXPANSION:
+${TENSILE_OUT}/1_BenchmarkProblems%.co: PROBLEM  = $(firstword $(subst /, ,$*))
+                                        OBJFILES = $(shell find ${TENSILE_OUT}/1_BenchmarkProblems/$(PROBLEM)/00_Final/source/build_tmp/SOURCE/assembly -name '*.o')
+${TENSILE_OUT}/1_BenchmarkProblems%.co: $$(OBJFILES)
+	@echo relinking $(PROBLEM) $(lastword $(subst /, ,$@))
+	@$(LDD) -target amdgcn-amd-amdhsa -Xlinker --build-id=sha1 ${LINK_ARGS} -o $@ $^
+
+%.o: %.s
+	@echo rebuilding $(lastword $(subst /, ,$@))
+	@$(AS) -x assembler -target amdgcn-amd-amdhsa -mcode-object-version=4 -mcpu=gfx90a ${WAVEFRONTSIZE} ${ASM_ARGS} -c -o $@ $^
```
Comment (jichangjichang) [inline tensilelite/Makefile]: why did you hardcode it as 90a?

**Comment by jichangjichang** [comment]
Is it better to create different Makefile for each problem folders by TensileCreatlibrary, which just follow the way to generate asm-new.sh?

**Comment by cliffxzx** [comment]
Hi @ellosel, it won't compile the asm (.s) file without the corresponding .co file. For example, if we add some random text to the asm file, which results in a compilation error, the .co file gets deleted. After that, no matter how I modify the asm file, it returns `make: Nothing to be done for 'all'`

> make co TENSILE_OUT=hhs_stable WAVEFRONTSIZE="-mno-wavefrontsize64"
>> rebuilding Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_MT64x64x3eUq8UMkgzpkCZgVyggFclXtojfMqCsdxbKRo6Nruhfk=.o
>> relinking Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_00 TensileLibrary_gfx1201.co

> make co TENSILE_OUT=hhs_stable WAVEFRONTSIZE="-mno-wavefrontsize64"
>> rebuilding Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_MT64x64x3eUq8UMkgzpkCZgVyggFclXtojfMqCsdxbKRo6Nruhfk=.o
>> hhs_stable/1_BenchmarkProblems/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_00/00_Final/source/build_tmp/SOURCE/assembly/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_MT64x64x3eUq8UMkgzpkCZgVyggFclXtojfMqCsdxbKRo6Nruhfk=.s:3725:1: error: undefined label 'label_GW_Beta_2'
>> s_cbranch_scc0 label_GW_Beta_2                     // Branch if Beta is not zero
>> ^
>> hhs_stable/1_BenchmarkProblems/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_00/00_Final/source/build_tmp/SOURCE/assembly/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_MT64x64x3eUq8UMkgzpkCZgVyggFclXtojfMqCsdxbKRo6Nruhfk=.s:3733:1: error: undefined label 'label_GW_B0_E1_1'
>> s_cbranch_scc1 label_GW_B0_E1_1                    // jump if edges required
>> ^
>> hhs_stable/1_BenchmarkProblems/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_00/00_Final/source/build_tmp/SOURCE/assembly/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_MT64x64x3eUq8UMkgzpkCZgVyggFclXtojfMqCsdxbKRo6Nruhfk=.s:3740:1: error: undefined label 'label_GW_B0_E1_1'
>> s_cbranch_scc1 label_GW_B0_E1_1                    // jump if edges required
>> ^
>> make: *** [Makefile:67: hhs_stable/1_BenchmarkProblems/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_00/00_Final/source/build_tmp/SOURCE/assembly/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_MT64x64x3eUq8UMkgzpkCZgVyggFclXtojfMqCsdxbKRo6Nruhfk=.o] Error 1

> make co TENSILE_OUT=hhs_stable WAVEFRONTSIZE="-mno-wavefrontsize64"
>> make: Nothing to be done for 'co'.

> make co TENSILE_OUT=hhs_stable WAVEFRONTSIZE="-mno-wavefrontsize64"
>> make: Nothing to be done for 'co'.

> make co TENSILE_OUT=hhs_stable WAVEFRONTSIZE="-mno-wavefrontsize64"
>> make: Nothing to be done for 'co'.

File: tensilelite/Makefile (hunk 5 above)
Comment (davidd-amd) [reply tensilelite/Makefile]: Good catch - I will correct this

File: tensilelite/Makefile (hunk 4 above)
Comment (davidd-amd) [reply tensilelite/Makefile]: I can default this any way you prefer.

**Comment by davidd-amd** [comment]
> Is it better to create different Makefile for each problem folders by TensileCreatlibrary, which just follow the way to generate asm-new.sh?

To goal is to simplify and reduce lines of code. Since we know the structure of the tensile output directory, we should have all of the information we need.

**Comment by davidd-amd** [comment]
> Hi @ellosel, it won't compile the asm (.s) file without the corresponding .co file. For example, if we add some random text to the asm file, which results in a compilation error, the .co file gets deleted. After that, no matter how I modify the asm file, it returns `make: Nothing to be done for 'all'`
> 
> > make co TENSILE_OUT=hhs_stable WAVEFRONTSIZE="-mno-wavefrontsize64"
> > > rebuilding Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_MT64x64x3eUq8UMkgzpkCZgVyggFclXtojfMqCsdxbKRo6Nruhfk=.o
> > > relinking Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_00 TensileLibrary_gfx1201.co
> 
> > make co TENSILE_OUT=hhs_stable WAVEFRONTSIZE="-mno-wavefrontsize64"
> > > rebuilding Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_MT64x64x3eUq8UMkgzpkCZgVyggFclXtojfMqCsdxbKRo6Nruhfk=.o
> > > hhs_stable/1_BenchmarkProblems/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_00/00_Final/source/build_tmp/SOURCE/assembly/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_MT64x64x3eUq8UMkgzpkCZgVyggFclXtojfMqCsdxbKRo6Nruhfk=.s:3725:1: error: undefined label 'label_GW_Beta_2'
> > > s_cbranch_scc0 label_GW_Beta_2                     // Branch if Beta is not zero
> > > ^
> > > hhs_stable/1_BenchmarkProblems/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_00/00_Final/source/build_tmp/SOURCE/assembly/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_MT64x64x3eUq8UMkgzpkCZgVyggFclXtojfMqCsdxbKRo6Nruhfk=.s:3733:1: error: undefined label 'label_GW_B0_E1_1'
> > > s_cbranch_scc1 label_GW_B0_E1_1                    // jump if edges required
> > > ^
> > > hhs_stable/1_BenchmarkProblems/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_00/00_Final/source/build_tmp/SOURCE/assembly/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_MT64x64x3eUq8UMkgzpkCZgVyggFclXtojfMqCsdxbKRo6Nruhfk=.s:3740:1: error: undefined label 'label_GW_B0_E1_1'
> > > s_cbranch_scc1 label_GW_B0_E1_1                    // jump if edges required
> > > ^
> > > make: *** [Makefile:67: hhs_stable/1_BenchmarkProblems/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_00/00_Final/source/build_tmp/SOURCE/assembly/Cijk_Alik_Bljk_HHS_BH_Bias_AS_UserArgs_MT64x64x3eUq8UMkgzpkCZgVyggFclXtojfMqCsdxbKRo6Nruhfk=.o] Error 1
> 
> > make co TENSILE_OUT=hhs_stable WAVEFRONTSIZE="-mno-wavefrontsize64"
> > > make: Nothing to be done for 'co'.
> 
> > make co TENSILE_OUT=hhs_stable WAVEFRONTSIZE="-mno-wavefrontsize64"
> > > make: Nothing to be done for 'co'.
> 
> > make co TENSILE_OUT=hhs_stable WAVEFRONTSIZE="-mno-wavefrontsize64"
> > > make: Nothing to be done for 'co'.

Correct the way it works is it interrogates the tensile-out directory for the co files in the library directory. If the co files aren't there it won't know to build them. However, I can't explain why it is deleting the co files. I thought that it would just replace existing co files. This should be easy enough to work around.

**Comment by davidd-amd** [comment]
@cliffxzx Please test the most recent commit. It should rebuild as expected. However, you will need to provide `ARCH` until I find a way to compute it. That is,

```
make co TENSILE_OUT=tensile-out ARCH=gfx942
```

File: tensilelite/Makefile (hunk 5 above)
Comment (davidd-amd) [reply tensilelite/Makefile]: For the time being, `ARCH` is a required input. I will try to find a way to compute it.

File: tensilelite/Makefile [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+# these can be customized as follows:
+#
+# make co TENSILE_OUT=tensile-out ARCH="gfx942:xnack-" WAVEFRONTSIZE="-mno-wavefrontsize64"
+#
+# In addition, we provide AS and ASM_ARGS, and LD LINK_ARGS as customization
+# points for the assemble and link step respectively.
+# ------------------------------------------------------------------------
+
+AS := /opt/rocm/bin/amdclang++
+LDD := /opt/rocm/bin/amdclang++
```
Comment (davidd-amd) [inline tensilelite/Makefile]: What about something like this:

```suggestion
ROCM_PATH ?= /opt/rocm
AS := $(ROCM_PATH)/bin/amdclang++
LDD := $(ROCM_PATH)/bin/amdclang++
```

Comment (davidd-amd) [reply tensilelite/Makefile]: @jichangjichang see most recent commit

**Comment by cliffxzx** [comment]
> @cliffxzx Please test the most recent commit. It should rebuild as expected. However, you will need to provide `ARCH` until I find a way to compute it. That is,
> 
> ```
> make co TENSILE_OUT=tensile-out ARCH=gfx942
> ```

Solved. Thanks 👍

File: tensilelite/Makefile [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+WAVEFRONTSIZE := -mwavefrontsize64
```
Comment (jichangjichang) [inline tensilelite/Makefile]: Could you provide a simple way for user to define the wavefront size? The original way of asm-new.sh is simpler.
ex:
- make  (no input, default is 64)
- make wave=64
- make wave=32
Then assign the corresponding setting into WAVEFRONTSIZE  in Makefile.
How do you think?

Comment (davidd-amd) [reply tensilelite/Makefile]: Sure looking at the asm-new.sh files there were two options: `-mwavefrontsize64` and `-mno-wavefrontsize64`

**Review by jichangjichang — APPROVED** [review]

**Review by bstefanuk — APPROVED** [review]

**Comment by ssuyuanchang** [comment]
[BF16_127x64x512_MT128x64x128.txt](https://github.com/user-attachments/files/18243806/BF16_127x64x512_MT128x64x128.txt)
Please rename BF16_127x64x512_MT128x64x128.txt to BF16_127x64x512_MT128x64x128.yaml.

Some errors happened ([make_co_error.log](https://github.com/user-attachments/files/18243842/make_co_error.log)) if you repeat the following steps:
1. ./tensilelite/Tensile/bin/Tensile BF16_127x64x512_MT128x64x128.yaml folder
2. touch folder/1_BenchmarkProblems/Cijk_Alik_Bljk_BBS_BH_UserArgs_00/00_Final/source/build_tmp/SOURCE/assembly/Cijk_Alik_Bljk_BBS_BH_UserArgs_MT128x64x128_MI16d2RqBrlcDzYPk5AX_h8Ym58MRf2qyNGRJs_dgfK9xu4\=.s
3. cd tensilelite
4. make co TENSILE_OUT=folder

**Comment by ssuyuanchang** [comment]
@ellosel Please help to fix this error

**Pull request merged**